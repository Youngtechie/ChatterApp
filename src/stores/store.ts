import { defineStore } from 'pinia'
import {ref} from 'vue'
import type { DocumentData } from 'firebase/firestore'

interface textarea{
  title: string;
  content: HTMLDivElement;
}

export const useChatterStore = defineStore('chatter', () => {
  const userId = ref('')
  const asReader = ref(true) 
  const section = 'personal' 
  const sidebar = ref(false) 
  const signedUser: DocumentData = {}
  const viwedPost: DocumentData = ref({})
  const viwedProfile: DocumentData = {}
  const coverImageFile = ref<File | null>(null)
  const viwedPosterFollowed = ref(false)
  const textarea: textarea[] = []

  const themeDetails = ref({
    theme: 'light',
    img: '/day.png',
    title: 'Day Mode'
  })

  // Theme-related functions
  function changeTheme() {
    // Updates the themeDetails based on the current theme
    themeDetails.value.img = themeDetails.value.theme === 'light' ? '/night.png' : '/day.png'
    themeDetails.value.title = themeDetails.value.theme === 'light' ? 'Dark Mode' : 'Light Mode'
    themeDetails.value.theme = themeDetails.value.theme === 'light' ? 'dark' : 'light'

    const app = document.querySelector('#app')
    if (app?.className === 'DayApp') {
      // Switches the class of the app element for theme switching
      document.querySelector('#app')?.setAttribute('class', 'NightApp')
    } else {
      app?.removeAttribute('class')
      document.querySelector('#app')?.setAttribute('class', 'DayApp')
    }
  }

  // Sidebar-related functions
  function toggleSidebar() {
    // Toggles the sidebar value
    sidebar.value = !sidebar.value
  }

  function toggleAccountType() {
    asReader.value = !asReader.value
    toggleSidebar()
  }

  const authenticated = false // Refactor: Use camelCase for variable names (authenticated)

  // User class
  class User {
    username = ''
    email = ''
    profilePicture = ''
    bio = ''
    dateOfBirth = ''
    location = ''
    followers: Record<string, any> = {
      total: 0,
      theFollowers: []
    }
    following: Record<string, any> = {
      total: 0,
      theFollowings: []
    }
    posts: string[] | null = []
    likes: Record<string, any> = {
      out: [],
      in: [],
      total: {
        out: 0,
        in: 0
      }
    }
    comments: Record<string, any> = {
      out: [],
      in: [],
      total: {
        out: 0,
        in: 0
      }
    }
    notifications: string[] = []
    settings: Record<string, any> = {
      privacySettings: {
        showEmail: true,
        showLastSeen: true,
        blockedUsers: []
      },
      securitySettings: {
        emailVerified: false
      }
    }
    interests: string[] = []
    activityLog: any[] = []
    savedPosts: string[] = []
    badges: string[] = []
    relationshipStatus = ''
    education: string[] = []
    workExperience: string[] = []
    lastActive = ''
    fullName = ''
    id = ''
    gender = ''
    isLogined = false
    blogName = ''
    contacts: string[] = []
    interactions: string[] = []
    bookmarks: string[] = []
    analyses: Record<string, any> = {
      totalComments: 0,
      totalLiked: 0,
      totalPosts: 0,
      totalPostsLikes: 0,
      totalPostsComments: 0,
      totalReposts: 0
    }

    constructor(email: string) {
      this.email = email
    }
  }

  class Post {
    posterId = ''
    postId = ''
    postTitle: string[] = []
    postContain = ""
    postRawContent = ""
    postTag = ''
    postTime = ""
    postLastModified = ""
    postLikes: Record<string, any> = {
        ids: [],
        total: 0
      }
    postComments: Record<string, any> = {
      details: [],
      total: 0,
      count: 0
    }
    postReposts = 0
    postViews = 0
    postBookmarks: string[] = []
    postMedia: string[] = []
    postSettings: Record<string, any> = {
      disableComments: false,
      allowReposts: true
    }
  }

  function createUser(email: string) {
    // Creates a new User instance with the provided email
    return new User(email)
  }

  function createPost(){
    return new Post
  }

  
  return {
    // Expose the necessary variables and functions from the store
    asReader,
    section,
    sidebar,
    signedUser,
    coverImageFile,
    themeDetails,
    authenticated,
    userId,
    viwedPost,
    viwedPosterFollowed,
    viwedProfile,
    textarea,
    changeTheme,
    toggleSidebar,
    createUser,
    toggleAccountType,
    createPost
  }
})
