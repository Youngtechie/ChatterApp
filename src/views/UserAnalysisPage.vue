<script setup lang="ts">
import { onUnmounted, type Ref, ref, nextTick, onMounted } from 'vue';
import { useRouter } from 'vue-router';
import { useChatterStore } from '@/stores/store';
import SignOut from '@/composables/useSignOut.vue';
import useUserDetails from '@/composables/useUserDetails.vue'
import useLoadingPage from "@/composables/useLoadingPage.vue";
import { Chart, registerables } from 'chart.js'

Chart.register(...registerables);

const router = useRouter()

const store = useChatterStore()

const isLoading: Ref<boolean> = ref(true)

onMounted(() => {
    useUserDetails()
})


let id = setTimeout(() => {
    if (store.signedUser.id === undefined && store.authenticated === false) {
        router.push('/home')
    }
    else if (store.authenticated === true) {
        if (store.signedUser.id !== undefined && store.signedUser.username === '') {
            const warningShow = document.getElementById('warningShow') as HTMLDivElement
            warningShow.style.display = 'flex'
            warningShow.textContent = 'Registration not complete... Logging you out'
            SignOut()
            id = setTimeout(() => {
                warningShow.style.display = 'none'
                store.authenticated = false
                router.push('/login')
            }, 2000)
        }
        else if (store.viwedProfile.username === undefined && store.viwedProfile.id === undefined) {
            nextTick(() => {
                const warning = document.getElementById('warningShow') as HTMLDivElement
                if (warning) {
                    warning.style.display = 'flex'
                    warning.textContent = 'Check your network connection and try reloading page'
                }
            })
        }
        isLoading.value = false
        createPie(store.signedUser.analyses.totalPosts, store.signedUser.analyses.totalComments, store.signedUser.analyses.totalLiked, store.signedUser.bookmarks.length)
    }
}, 5000)

onUnmounted(() => {
    clearTimeout(id)
    const warning = document.getElementById('warningShow') as HTMLDivElement
    if (warning) {
        warning.style.display = 'none'
    }
})

function back() {
    router.go(-1)
}

function createPie(totalPosts: any, totalComments: any, totalLiked: any, totalBookmarks: any) {

    var data = {
        labels: ['Posts', 'Total Comments', 'Posts Like', 'Bookmarks'],
        datasets: [{
            data: [totalPosts, totalComments, totalLiked, totalBookmarks],
            backgroundColor: ['#FF6384', '#36A2EB', '#FFCE56', '#4BC0C0'],
            hoverBackgroundColor: ['#FF6384', '#36A2EB', '#FFCE56', '#4BC0C0']
        }]
    }

    nextTick(() => {
        var ctx = (document.getElementById('pieChart') as HTMLCanvasElement).getContext('2d');
        new Chart(ctx as CanvasRenderingContext2D, {
            type: 'pie',
            data: data,
            options: {
                responsive: false, // Disable responsiveness
            }
        });
    })
}

</script>
<template>
    <useLoadingPage v-if="isLoading" action-name="Loading Analysis..." />
    <main v-else class="analysisPage">
        <div id="warningShow"></div>
        <button @click.prevent="back">Back</button>
        <h1>{{ store.signedUser.username }} Analyses</h1>
        <section>
            <p><span>Total Posts:</span> {{ store.signedUser.analyses.totalPosts }}</p>
            <p><span>Total Comments Made:</span> {{ store.signedUser.analyses.totalComments }}</p>
            <p><span>Total Posts Liked:</span> {{ store.signedUser.analyses.totalLiked }}</p>
            <p><span>Bookmarks: </span> {{ store.signedUser.bookmarks.length }} </p>
        </section>
        <canvas id="pieChart"></canvas>
    </main>
</template>
<style scoped>
button {
    position: absolute;
    top: 10px;
    left: 10px;
    padding: 5px;
    cursor: pointer;
    border-radius: 10px;
}

button:hover{
    background-color: blue;
    color: white;
}

.analysisPage {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    background-color: #f5f5f5;
    color: black;
    height: 100vh;
    position: relative;
}

.analysisPage h1 {
    margin-bottom: 10px;
    text-align: center;
}

.analysisPage section {
    margin-top: 20px;
}

.analysisPage h2 {
    font-size: 20px;
    margin-bottom: 10px;
}

.analysisPage p {
    margin-bottom: 5px;
}

.analysisPage p span {
    font-weight: bold;
    font-size: larger;
}
</style>