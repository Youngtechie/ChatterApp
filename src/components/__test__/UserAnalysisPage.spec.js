import {mount} from '@vue/test-utils'
import {describe, expect, it, beforeEach, vi} from 'vitest'
import UserAnalysisPage from '@/views/UserAnalysisPage.vue'
import {setActivePinia, createPinia} from 'pinia'
import {useChatterStore} from '@/stores/store'

let store;

beforeEach(() => {
    setActivePinia(createPinia());
    
    store = useChatterStore();
})

describe('UserAnalysisPage rendered correctly', () => {
    it('should render correctly', () => {
        const wrapper = mount(UserAnalysisPage);
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('should have a signedUser', async () => {
        new Promise((resolve) => {
            setTimeout(resolve, 5000);
        }).then(() => {
            expect(store.signedUser.length).toBeGreaterThan(0);
            const wrapper = mount(UserAnalysisPage);
            const createPie = vi.fn(wrapper.vm.createPie);
            createPie(1,1,3,4);
            expect(createPie).toHaveBeenCalled(2);
            expect(wrapper.find('#pieChart')).childNodes.length.toBeGreaterThan(0); 
        });
    })
})