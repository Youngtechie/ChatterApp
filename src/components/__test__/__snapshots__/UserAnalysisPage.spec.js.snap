// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserAnalysisPage rendered correctly > should render correctly 1`] = `
"<div data-v-2c6dcbed=\\"\\" class=\\"loading-page\\">
  <div class=\\"spinner\\"></div>
  <h1>Loading Analysis....</h1>
</div>"
`;
